File: mobile/lib/models/product_variant.dart
/// Ось вариативности товара («Размер», «Цвет») с её значениями.
class ProductOption {
  final String name;
  final int position;
  final List<String> values;

  const ProductOption({
    required this.name,
    required this.position,
    required this.values,
  });

  factory ProductOption.fromJson(Map<String, dynamic> json) => ProductOption(
    name: (json['name'] as String?)?.trim() ?? '',
    position: (json['position'] as num?)?.toInt() ?? 1,
    values: ((json['values'] as List<dynamic>?) ?? const [])
        .map((v) => (v as Map<String, dynamic>)['value']?.toString() ?? '')
        .where((v) => v.isNotEmpty)
        .toList(),
  );
}

/// Заведённая комбинация опций — свой остаток, цена, SKU, фото.
/// optionValues — значения по позициям опций (напр. ["M", "Чёрный"]).
class ProductVariant {
  final String id;
  final String? sku;

  /// Цена варианта в копейках; null → цена товара.
  final int? priceKopecks;
  final int stockQuantity;
  final bool allowBackorder;
  final String? imageUrl;
  final List<String> optionValues;
  final bool isActive;

  const ProductVariant({
    required this.id,
    this.sku,
    this.priceKopecks,
    required this.stockQuantity,
    required this.allowBackorder,
    this.imageUrl,
    required this.optionValues,
    required this.isActive,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) => ProductVariant(
    id: json['id'] as String,
    sku: json['sku'] as String?,
    priceKopecks: (json['price'] as num?)?.toInt(),
    stockQuantity: (json['stock_quantity'] as num?)?.toInt() ?? 0,
    allowBackorder: json['allow_backorder'] as bool? ?? false,
    imageUrl: json['image_url'] as String?,
    optionValues: ((json['option_values'] as List<dynamic>?) ?? const [])
        .map((v) => v.toString())
        .toList(),
    isActive: json['is_active'] as bool? ?? true,
  );

  bool get inStock => isActive && (stockQuantity > 0 || allowBackorder);

  int effectivePriceKopecks(int productPriceKopecks) =>
      priceKopecks ?? productPriceKopecks;
}

File: app/Models/ProductVariant.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;

/**
 * Реально заведённая комбинация опций товара — своя цена/остаток/SKU/фото.
 * option_values — массив значений строками, позиционно к ProductOption.position
 * (напр. ["M", "Чёрный"]). Цена null → наследует products.price.
 */
class ProductVariant extends Model
{
    use HasUuids;

    protected $table = 'product_variants';

    protected $fillable = [
        'product_id',
        'sku',
        'price',
        'stock_quantity',
        'allow_backorder',
        'image_url',
        'option_values',
        'is_active',
        'position',
    ];

    protected $casts = [
        'price'           => 'integer',
        'stock_quantity'  => 'integer',
        'allow_backorder' => 'boolean',
        'option_values'   => 'array',
        'is_active'       => 'boolean',
        'position'        => 'integer',
        'created_at'      => 'datetime',
        'updated_at'      => 'datetime',
    ];

    public function product()
    {
        return $this->belongsTo(Product::class, 'product_id');
    }

    /** Цена варианта в копейках — своя, иначе цена товара. */
    public function effectivePrice(int $productPrice): int
    {
        return $this->price ?? $productPrice;
    }

    /** «S · Чёрный» — из значений, без имён опций (для короткой подписи). */
    public function shortLabel(): string
    {
        return implode(' · ', array_filter((array) $this->option_values, fn ($v) => $v !== null && $v !== ''));
    }

    /**
     * «Размер: M · Цвет: Чёрный» — снимок для order_items.variant_label.
     * @param array<int,string> $optionNames имена опций по позиции (0-based)
     */
    public function fullLabel(array $optionNames): string
    {
        $parts = [];
        foreach ((array) $this->option_values as $i => $val) {
            if ($val === null || $val === '') {
                continue;
            }
            $name = $optionNames[$i] ?? null;
            $parts[] = $name ? "{$name}: {$val}" : $val;
        }
        return implode(' · ', $parts);
    }

    public function inStock(): bool
    {
        return $this->is_active && ($this->stock_quantity > 0 || $this->allow_backorder);
    }
}
